c/recovery.c

```diff
@@ -947,20 +1095,46 @@ int recoveryEstablishRouter(int flags) {
     previousESPIEToken = setDummyESPIE();
   }
 
+  int rc = RC_RCV_OK;
+
   /* init recovery block */
-  RecoveryContext *context = (RecoveryContext *)storageObtain(sizeof(RecoveryContext));
+  RecoveryContext *context = NULL;
+  if (userContext == NULL) {
+    context = (RecoveryContext *)storageObtain(sizeof(RecoveryContext));
+    if (context == NULL) {
+      rc = RC_RCV_ALLOC_FAILED;
+      goto failure;
+    }
+  } else {
+    flags |= RCVR_ROUTER_FLAG_USER_CONTEXT;
+    context = userContext;
+  }
+
   int setRC = setRecoveryContext(context);
   if (setRC != RC_RCV_OK) {
-    storageRelease((char *)context, sizeof(RecoveryContext));
-    context = NULL;
-    return RC_RCV_CONTEXT_NOT_SET;
+    rc = RC_RCV_CONTEXT_NOT_SET;
+    goto failure;
   }
 
   memset(context, 0, sizeof(RecoveryContext));
   memcpy(context->eyecatcher, "RSRCVCTX", sizeof(context->eyecatcher));
-  context->flags = flags;
   context->previousESPIEToken = previousESPIEToken;
 
+#ifdef RCVR_CPOOL_STATES
+  if (userStatePool != RCVR_STATE_POOL_NULL) {
+    context->statePool = userStatePool;
+    flags |= RCVR_ROUTER_FLAG_USER_STATE_POOL;
+  } else {
+    context->statePool = makeStatePool(RCVR_STATE_POOL_PRIMARY_COUNT,
```

Comment:
I'm unsure of what kind of state pool this is. I know from looking at the code it is a recovery state pool so I would rename the function to makeRecorveryStatePool.

Reply:
👍

Reply:
Done.